test/functional/testlib.bash

```diff
@@ -0,0 +1,502 @@
+#!bin/bash
+
+FUNC_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+TEST_FILENAME=$(basename "$BATS_TEST_FILENAME")
+TEST_NAME=${TEST_FILENAME%.bats}
+export TEST_NAME
+export FUNC_DIR
+export SWUPD_DIR="$FUNC_DIR/../.."
+export SWUPD="$SWUPD_DIR/swupd"
+
+
+generate_random_content() { 
+
+	local bottom_range=5
+	local top_range=100
+	local range=$((top_range - bottom_range + 1))
+	local number_of_lines=$((RANDOM%range))  
+	< /dev/urandom tr -dc 'a-zA-Z0-9-_!@#$%^&*()_+{}|:<>?=' | fold -w 100 | head -n $number_of_lines
+
+}
+
+generate_random_name() { 
+
+	local prefix=${1:-test-}
+	local uuid
+	
+	# generate random 8 character alphanumeric string (lowercase only)
+	uuid=$(< /dev/urandom tr -dc 'a-f0-9' | fold -w 8 | head -n 1)
+	echo "$prefix$uuid"
+
+}
+
+validate_path() { 
+
+	local path=$1
+	if [ -z "$path" ] || [ ! -d "$path" ]; then
+		echo "Please provide a valid path"
+		return 1
+	fi
+
+}
+
+validate_item() { 
+
+	local vfile=$1
+	if [ -z "$vfile" ] || [ ! -e "$vfile" ]; then
+		echo "Please provide a valid file"
```

Comment:
Should abstract out the printing of the error message to another function. Then it will be easy to fix all of them to output to stderr, rather than stdout .

Reply:
I did something similar to the validate_param function here so the program is stopped if the file doesn't exists.

```bash
validate_item() { 

	local vfile=$1
	local param
	if [ -z "$vfile" ] || [ ! -e "$vfile" ]; then
		# since the lib could be sourced and run from an interactive shell
		# not only from a script, we cannot use exit since it would terminate
		# the shell, so we are using parameter expansion with a fake parameter
		# "param" to force an error
		: ${param:?"Please provide a valid file"}
	fi

}
```

Reply:
all messages from stdout and stderr are redirected to stderr if a test fails in bats anyway